File: construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        def solver(bound=None):
            if not inorder or inorder[0]==bound: return None
            root=TreeNode(preorder.pop(0))
            root.left=solver(root.val)
            inorder.pop(0)
            root.right=solver(bound)
            return root
        
        return solver()

File: construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return solve(preorder,inorder,Integer.MIN_VALUE);    
    }
    
    int pre=0,in=0;
    
    TreeNode solve(int[]  preorder,int[] inorder,int end) {
        if(pre==preorder.length)
            return null;
        
        if(inorder[in]==end) {
            in++;
            return null;
        }
        TreeNode node=new TreeNode(preorder[pre++]);
        node.left=solve(preorder,inorder,node.val);
        node.right=solve(preorder,inorder,end);
        return node;
    }
}